### ISSUE TITLE
Fix read-line so end-of-file can be detected

### ISSUE THREAD

**Issue by hanche** [elves/elvish#1509]
Currently `read-line` returns an empty string when invoked at end-of-file. This is the same you get for an empty line, since the trailing LF or CRLF is stripped. And so there is no way to detect an eof condition.

Suggestion: Let `read-line` return `$nil` when invoked at eof. (Of course, it should still return a string if the current line is nonempty and end at eof.)

Here is a workaround:

```
fn read-line {
  use str
  str:trim-suffix (str:trim-suffix (read-upto "\n") "\n") "\r")
}
```

(Not sure if this does the right thing if the line ends in CR followed by eof, but that does not seem important.)

**Comment by xiaq** [comment]
Hmm, this will make one-off invocations of `read-line` more complex because they now need to deal with `$nil`.

What's the scenario you're using `read-line` in where the current behavior is a problem? If you're trying to consume the input line-by-line, you should use `from-lines`.

Also, observe this:

* `a<EOF>` -> return `a`
* `<EOF>` -> return `$nil`

**Comment by hanche** [comment]
One scenario is reading input from a long-running process, processing it a line at a time.

As for complicating other uses of `read-line`, one could make it an option. Say `&eof` with the empty string as default value. to preserve the current behaviour. I could then use `read-line &eof=$nil` to get the new behaviour.

You final observations above show intended behaviour, as far as I am concerned. A user who want to distinguish whether the final line of the file ends in a newline or eof would have to use `read-upto "\n"` instead.

**Comment by xiaq** [comment]
I like the `&eof` option idea.

**Comment by xiaq** [comment]
Actually, I am wondering if there is a more general design. The `&eof` option only works when there is nothing to be read; it doesn't allow distinguishing when the remaining text is `"foo\n"` vs just `"foo"`.

**Comment by krader1961** [comment]
> Actually, I am wondering if there is a more general design. The &eof option only works when there is nothing to be read; it doesn't allow distinguishing when the remaining text is "foo\n" vs just "foo".

I'm pretty sure I pointed that out when the introduction of `read-line` was being discussed.

I don't think the `$eof` option (assuming we agree it should be introduced) should take a value since it invites abuse. What if someone writes `read-line &eof=""`? Thus shooting themselves in the foot since their code couldn't distinguish an empty line from EOF. In fact, I'm not convinced an option is even warranted. Yes, changing the existing behavior to the equivalent of `read-line &eof=$nil` is a breaking change. But how many uses of `read-line` are already broken because they can't recognize EOF? In my local config I have just two uses of `read-line`:

```
.local/share/elvish/lib/github.com/zzamboni/elvish-modules/tty.org:  pass = (tty:with-stty [-echo] { read-line })
.local/share/elvish/lib/github.com/zzamboni/elvish-modules/util.org:    var fname = (read-line)
```

The first will never return `$nil` since it is reading from the tty and if it is close the shell will exit. The second also appears to never be able to return `$nil` since it is reading from the tty. Yes, there are undoubtedly other uses of `read-line` but it seems to me they are already broken since they can't detect EOF.

Having written the above I'm inclined to think `read-line` should throw an exception on EOF. Perhaps, the user should have to use a `try` block if they want to handle calling `read-line` at EOF. The question is whether just `"foo"`, without a trailing line ending at EOF, should return `"foo"` or throw the EOF exception. I'm inclined to think it should return `"foo"` with the next `read-line` throwing the EOF exception.

**Comment by xiaq** [comment]
> The question is whether just "foo", without a trailing line ending at EOF, should return "foo" or throw the EOF exception. I'm inclined to think it should return "foo" with the next read-line throwing the EOF exception.

Neither option allows distinguishing `foo` followed directly by EOF or by a newline then EOF?

**Comment by hanche** [comment]
> The first will never return `$nil` since it is reading from the tty and if it is close the shell will exit.

No, it won't exit. The behaviour of a tty (on unix – I have no idea what happens on windows) is somewhat peculiar: All that happens when you type ^D with the tty in its normal (“sane”) state, is that the wating `read` call will immediately return whatever is in the buffer. Just try it: Run `read-line` or `read-upto "\n"` from the interactive prompt, type some characters, and end it with ^D. You can do it repeatedly, including with an empty input; you always get what you typed back as a string value, and elvish will most definitely not exit. In a sense, you can get multiple “eof” events on the tty.

In C, if you have a high level stream attached to the tty, it will report EOF if the `read` call returns an empty string, as happens when you hit ^D at the start of a line – or if you hit ^D twice elsewhere, as the first one will cause the buffer to be emptied. I assume other languages, such as go, behave in the same fashion.

**Comment by hanche** [comment]
> > The question is whether just "foo", without a trailing line ending at EOF, should return "foo" or throw the EOF exception. I'm inclined to think it should return "foo" with the next read-line throwing the EOF exception.
> 
> Neither option allows distinguishing `foo` followed directly by EOF or by a newline then EOF?

A third option is to do both: Since elvish functions don't really “return” values, but rather puts them to the output stream, you can both put the value read and then throw an exception. One can certainly question whether this is good idea, as it seems a bit tricky to use in practice, but it is definitely a possibility that should not be dismissed out of hand.

**Comment by krader1961** [comment]
> Neither option allows distinguishing foo followed directly by EOF or by a newline then EOF?

Yes, but that's true right now:

```
> { echo one; print two } | { read-line; read-line; read-line }
▶ one
▶ two
▶ ''
```

I currently inclined to add a new option, `&retain-newline` (or some such), and having EOF raise an exception. Alternatively, make a breaking change so that retaining the newline is the default behavior. Either way the user can then do `str:has-suffix  $line "\n"` to determine if the line terminator is present, and`str:trim-right $line "\r\n"` to remove it.
